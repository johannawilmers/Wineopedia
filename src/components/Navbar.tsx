import { Link } from "react-router-dom";
import "../styles/navbar.css";

export function Navbar() {
  return (
    <nav className="navbar">
      <div className="navbarContent">
        <Link to={""}>HOME </Link>
        <Link to={"myfavorites"}> MY FAVORITES</Link>
      </div>
    </nav>
  );
}
