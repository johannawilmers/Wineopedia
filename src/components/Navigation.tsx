import { Route, Routes } from "react-router-dom";
import Favorites from "../pages/Favorites";
import Homepage from "../pages/Homepage";
import Product from "../pages/Product";

/**
 * This component is used to display a list of wine cards
 */
function Navigation() {
  return (
    <div>
      <Routes>
        <Route path="/myfavorites" Component={Favorites}></Route>
        <Route path="/" Component={Homepage}></Route>
        <Route path="/product/:id" Component={Product} />
      </Routes>
    </div>
  );
}

export default Navigation;
