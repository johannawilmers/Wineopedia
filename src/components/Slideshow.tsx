import React from "react";
import { Wine } from "../utils/types";
import { useFavorites } from "../hooks/useFavorites";
import { useFilter } from "../hooks/useFilter";
import "../styles/slideshow.css";
import "../styles/listView.css";

interface WineSlideShow {
  wines: Wine[];
  currentIndex: number;
  activeButton: string;
  setCurrentIndex: (index: number) => void;
}

/**
 * This component is used to display a slideshow of wines, showing one wine at a time
 * @param {WineSlideShow} wines
 */
const Slideshow: React.FunctionComponent<WineSlideShow> = ({ wines }) => {
  const { favorites, toggleFavorite } = useFavorites();
  const { currentIndex, setCurrentIndex } = useFilter();
  if (wines.length === 0) {
    return (
      <div id="errorNoWine">
        <p>No products in this category. Add your favorites in homepage</p>
      </div>
    );
  }
  const currentWine = wines[currentIndex]?.basic.productId
    ? wines[currentIndex] || {}
    : wines[0];
  const isFavorite =
    currentWine.basic && favorites.includes(currentWine.basic.productId);

  const getPreviousElement = () => {
    currentIndex !== 0 ? setCurrentIndex(currentIndex - 1) : setCurrentIndex(0);
  };

  const getNextElement = () => {
    currentIndex < wines.length - 1
      ? setCurrentIndex(currentIndex + 1)
      : setCurrentIndex(0);
  };

  return (
    <div>
      <div className="slideShow">
        <button
          className="buttonLeft"
          aria-label="button-ClickLeft"
          onClick={getPreviousElement}
        />
        <div className="wineSlide">
          <h2 className="slideshowProductName">
            {currentWine.basic?.productShortName ||
              "Product Name Not Available"}
          </h2>
          {currentWine.basic && (
            <img
              className="wineImage"
              src={`https://bilder.vinmonopolet.no/cache/300x300-0/${currentWine.basic.productId}-1.jpg`}
              alt={currentWine.basic.productShortName}
            />
          )}
        </div>
        <button
          className="buttonRight"
          aria-label="button-ClickRight"
          onClick={getNextElement}
        />
      </div>
      <div className="wineInformation">
        <p>
          <b> ProductID: </b> {currentWine.basic?.productId || "N/A"}
        </p>
        <p>
          <b> Last Changed at: </b> {currentWine.lastChanged?.date}{" "}
          {currentWine.lastChanged?.time}
        </p>
        <div className="buttonContainer">
          <button
            onClick={() => toggleFavorite(currentWine.basic?.productId)}
            className="favoriteButton"
          >
            {isFavorite ? "Remove from Favorites" : "Add to Favorites"}
          </button>
        </div>
      </div>
    </div>
  );
};

export default Slideshow;
