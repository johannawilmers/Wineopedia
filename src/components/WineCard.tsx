import React from "react";
import { Wine } from "../utils/types";
import "../styles/wineCard.css";

interface WineCardProps {
  wine: Wine;
}

/**
 * This component is used to display a single wine card
 * @param {ineCardProps} wine
 */
const WineCard: React.FunctionComponent<WineCardProps> = ({ wine }) => {
  return (
    <a href={`/project1/product/${wine.basic.productId}`} className="wineCard">
      <div>
        <img
          src={`https://bilder.vinmonopolet.no/cache/300x300-0/${wine.basic.productId}-1.jpg`}
          alt={wine.basic.productShortName}
        />
      </div>
      <h2 className="wineName">{wine.basic.productShortName}</h2>
    </a>
  );
};

export default WineCard;
