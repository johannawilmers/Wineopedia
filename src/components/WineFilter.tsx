import { useFilter } from "../hooks/useFilter";

interface WineFilterProps {
  setCurrentWineString: (wineType: string) => void;
}

/**
 * This component is used to filter on red and white wines
 */
export const WineFilter = (props: WineFilterProps) => {
  const { activeButton, setActiveButton, setCurrentIndex } = useFilter();

  const handleFilterClick = (wineType: string) => {
    props.setCurrentWineString(wineType);
    setActiveButton(wineType);
    setCurrentIndex(0);
  };

  return (
    <div className="buttonContainer">
      <button
        className={`filterButton ${activeButton === "rød" ? "active" : ""}`}
        onClick={() => handleFilterClick("rød")}
      >
        Red
      </button>
      <button
        className={`filterButton ${activeButton === "" ? "active" : ""}`}
        onClick={() => handleFilterClick("")}
      >
        All wines
      </button>
      <button
        className={`filterButton ${activeButton === "hvit" ? "active" : ""}`}
        onClick={() => handleFilterClick("hvit")}
      >
        White
      </button>
    </div>
  );
};
