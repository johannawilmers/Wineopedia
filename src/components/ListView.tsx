import React, { useState } from "react";
import { Wine } from "../utils/types";
import "../styles/listView.css";
import WineCard from "./WineCard";
import { useFavorites } from "../hooks/useFavorites";

interface WineListView {
  wines: Wine[];
}

/**
 * This component is used to display a list of wine cards
 * @param {WineListView} wines
 */
const ListView: React.FunctionComponent<WineListView> = ({ wines }) => {
  const [visibleItems, setVisibleItems] = useState(6);
  const { favorites, toggleFavorite } = useFavorites();

  const showMore = () => {
    setVisibleItems((prevVisibleItems) => prevVisibleItems + 6);
  };

  return (
    <div>
      <div className="listview-container">
        {wines.slice(0, visibleItems).map((wine) => (
          <div key={wine.basic.productId} className="wine-card-wrapper">
            <WineCard wine={wine} />
            <button
              onClick={() => toggleFavorite(wine.basic?.productId)}
              className="favoriteButton"
            >
              {wine.basic && favorites.includes(wine.basic.productId)
                ? "Remove from Favorites"
                : "Add to Favorites"}
            </button>
          </div>
        ))}
      </div>
      <div id="buttonDiv" data-testid="favoriteButton">
        {visibleItems < wines.length && (
          <button className="show-more-button" onClick={showMore}>
            Load more wines
          </button>
        )}
      </div>
    </div>
  );
};

export default ListView;
