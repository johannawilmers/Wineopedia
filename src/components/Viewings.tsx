import Slideshow from "./Slideshow";
import { Wine } from "../utils/types";
import ListView from "./ListView";
import { useEffect, useState } from "react";
import { WineFilter } from "./WineFilter";
import { useViewing } from "../hooks/useViewing";
import { useFilter } from "../hooks/useFilter";

/**
 * This component is used to display either a slideshow or listview
 */
export const Viewings = () => {
  const { isSlideShow, setIsSlideshow } = useViewing();
  const [wineString, setWineString] = useState("");
  const { currentIndex, setCurrentIndex } = useFilter();
  const { activeButton } = useFilter();

  useEffect(() => {
    setWineString(activeButton);
  }, [activeButton]);

  const setViewing = (map: Wine[]) => {
    if (isSlideShow) {
      return (
        <div>
          <WineFilter setCurrentWineString={setWineString} />
          <Slideshow
            currentIndex={currentIndex}
            setCurrentIndex={setCurrentIndex}
            wines={map.filter((w) =>
              w.basic.productShortName.toLowerCase().includes(wineString),
            )}
            activeButton={""}
          />
        </div>
      );
    }
    return (
      <div>
        <WineFilter setCurrentWineString={setWineString} />
        <h2 id="NBheader">
          Some of the wines may have the same name, but they differ in sizing
          and packing. Click on a wine to get more information!
        </h2>
        <hr></hr>
        <ListView
          wines={map.filter((w) =>
            w.basic.productShortName.toLowerCase().includes(wineString),
          )}
        />
      </div>
    );
  };

  const changeViewing = () => {
    setIsSlideshow((current) => !current);
  };

  return { setViewing, changeViewing, isSlideShow };
};
