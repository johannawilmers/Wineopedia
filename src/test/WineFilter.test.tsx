import {
  render,
  screen,
  fireEvent,
  RenderResult,
} from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import "@testing-library/jest-dom/extend-expect";
import ListView from "../components/ListView";
import { WineFilter } from "../components/WineFilter";
import Slideshow from "../components/Slideshow";

const allWines = [
  {
    basic: {
      productId: "133501",
      productShortName: "Tissot Vin Jaune La Vasée",
      favorite: false,
    },
    lastChanged: {
      date: "2021-04-16",
      time: "06:19:05",
    },
  },
  {
    basic: {
      productId: "162701",
      productShortName: "Niepoort Bioma Vintage",
      favorite: false,
    },
    lastChanged: {
      date: "2021-03-15",
      time: "22:57:42",
    },
  },
];

const redWines = [
  {
    basic: {
      productId: "257201",
      productShortName: "Smalhans rødvin",
      favorite: false,
    },
    lastChanged: {
      date: "2021-03-15",
      time: "22:57:42",
    },
  },
  {
    basic: {
      productId: "650401",
      productShortName: "Grevinnen & Hovmesteren Rødvin",
      favorite: false,
    },
    lastChanged: {
      date: "2021-03-15",
      time: "22:57:42",
    },
  },
];

const whiteWines = [
  {
    basic: {
      productId: "201001",
      productShortName: "Smalhans hvitvin",
      favorite: false,
    },
    lastChanged: {
      date: "2021-03-15",
      time: "22:57:42",
    },
  },
  {
    basic: {
      productId: "6436006",
      productShortName: "Hvit Hagevin Arne & Carlos",
      favorite: false,
    },
    lastChanged: {
      date: "2021-03-15",
      time: "22:57:42",
    },
  },
];

describe("renders filter component in listView", () => {
  let wrapper: RenderResult;

  it("test the filter for all wines", () => {
    wrapper = render(
      <BrowserRouter>
        <div>
          <WineFilter setCurrentWineString={() => {}} />
          <ListView wines={allWines} />
        </div>
      </BrowserRouter>,
    );

    const wine1 = screen.getByText("Tissot Vin Jaune La Vasée");
    const wine2 = screen.getByText("Niepoort Bioma Vintage");
    expect(wine1 && wine2).toBeInTheDocument();

    expect(wine1).toMatchSnapshot();
    expect(wine2).toMatchSnapshot();
  });

  it("test the filter for red wines", () => {
    wrapper = render(
      <BrowserRouter>
        <div>
          <WineFilter setCurrentWineString={() => {}} />
          <ListView wines={redWines} />
        </div>
      </BrowserRouter>,
    );
    const redbutton = wrapper.getByRole("button", { name: /Red/i });
    expect(redbutton).toBeInTheDocument();
    fireEvent.click(redbutton);
    const wine1 = screen.getByText("Smalhans rødvin");
    const wine2 = screen.getByText("Grevinnen & Hovmesteren Rødvin");
    expect(wine1 && wine2).toBeInTheDocument();
  });

  it("test the filter for white wines", () => {
    wrapper = render(
      <BrowserRouter>
        <div>
          <WineFilter setCurrentWineString={() => {}} />
          <ListView wines={whiteWines} />
        </div>
      </BrowserRouter>,
    );
    const button = screen.getByRole("button", { name: /White/i });
    fireEvent.click(button);
    const wine = screen.getByRole("heading", { name: /Smalhans hvitvin/i });
    expect(wine).toBeInTheDocument();
  });

  it("dobble check the that the filter for all wines works", () => {
    wrapper = render(
      <BrowserRouter>
        <div>
          <WineFilter setCurrentWineString={() => {}} />
          <ListView wines={redWines} />
        </div>
      </BrowserRouter>,
    );
    const button = screen.getByRole("button", { name: /Red/i });
    fireEvent.click(button);
    const wine = screen.queryByText("Smalhans hvitvin");
    expect(wine).not.toBeInTheDocument();
  });
});

describe("renders slideshow component with filter", () => {
  it("test the filter for all wines", () => {
    render(
      <BrowserRouter>
        <div>
          <WineFilter setCurrentWineString={() => {}} />
          <Slideshow
            wines={allWines}
            currentIndex={0}
            activeButton={"All wines"}
            setCurrentIndex={function (): void {
              throw new Error("Function not implemented.");
            }}
          />
        </div>
      </BrowserRouter>,
    );
    const wine1 = screen.getByText("Tissot Vin Jaune La Vasée");
    expect(wine1).toBeInTheDocument();

    const button = screen.getByRole("button", { name: /button-ClickRight/i });
    fireEvent.click(button);
    const wine2 = screen.getByText("Niepoort Bioma Vintage");
    expect(wine2).toBeInTheDocument();
  });

  it("test the filter for white wines", () => {
    render(
      <BrowserRouter>
        <div>
          <WineFilter setCurrentWineString={() => {}} />
          <Slideshow
            wines={whiteWines}
            currentIndex={0}
            activeButton={"All wines"}
            setCurrentIndex={function (): void {
              throw new Error("Function not implemented.");
            }}
          />
        </div>
      </BrowserRouter>,
    );

    const button = screen.getByRole("button", { name: /White/i });
    fireEvent.click(button);
    const wine = screen.getByText("Smalhans hvitvin");
    expect(wine).toBeInTheDocument();
  });

  it("test that you load 6 wines at the time", () => {
    const loadMoreWines = [
      {
        basic: {
          productId: "1",
          productShortName: "Vin1",
          favorite: false,
        },
        lastChanged: {
          date: "2021-04-16",
          time: "06:19:05",
        },
      },
      {
        basic: {
          productId: "2",
          productShortName: "Vin2",
          favorite: false,
        },
        lastChanged: {
          date: "2021-03-15",
          time: "22:57:42",
        },
      },
      {
        basic: {
          productId: "3",
          productShortName: "Vin3",
          favorite: false,
        },
        lastChanged: {
          date: "2021-04-16",
          time: "06:19:05",
        },
      },
      {
        basic: {
          productId: "4",
          productShortName: "Vin4",
          favorite: false,
        },
        lastChanged: {
          date: "2021-03-15",
          time: "22:57:42",
        },
      },
      {
        basic: {
          productId: "5",
          productShortName: "Vin5",
          favorite: false,
        },
        lastChanged: {
          date: "2021-04-16",
          time: "06:19:05",
        },
      },
      {
        basic: {
          productId: "6",
          productShortName: "Vin6",
          favorite: false,
        },
        lastChanged: {
          date: "2021-03-15",
          time: "22:57:42",
        },
      },
      {
        basic: {
          productId: "7",
          productShortName: "Vin7",
          favorite: false,
        },
        lastChanged: {
          date: "2021-04-16",
          time: "06:19:05",
        },
      },
      {
        basic: {
          productId: "8",
          productShortName: "Vin8",
          favorite: false,
        },
        lastChanged: {
          date: "2021-03-15",
          time: "22:57:42",
        },
      },
    ];

    render(
      <BrowserRouter>
        <div>
          <WineFilter setCurrentWineString={() => {}} />
          <ListView wines={loadMoreWines} />
        </div>
      </BrowserRouter>,
    );

    expect(screen.getByText("Vin1")).toBeInTheDocument();
    expect(screen.getByText("Vin2")).toBeInTheDocument();
    expect(screen.getByText("Vin3")).toBeInTheDocument();
    expect(screen.getByText("Vin4")).toBeInTheDocument();
    expect(screen.getByText("Vin5")).toBeInTheDocument();
    expect(screen.getByText("Vin6")).toBeInTheDocument();

    // test that the load more button loads more wines when you click the button
    const loadMoreWinesButton = screen.getByRole("button", {
      name: /Load more wines/i,
    });
    fireEvent.click(loadMoreWinesButton);

    expect(screen.getByText("Vin7")).toBeInTheDocument();
    expect(screen.getByText("Vin8")).toBeInTheDocument();
  });

  it("test that the favorite button changes text when you click it", () => {
    render(
      <BrowserRouter>
        <div>
          <WineFilter setCurrentWineString={() => {}} />
          <ListView wines={allWines} />
        </div>
      </BrowserRouter>,
    );
    const favoriteButtons = screen.getAllByText("Add to Favorites");
    fireEvent.click(favoriteButtons[0]);
    expect(favoriteButtons[0]).toHaveTextContent(/Remove from Favorites/i);
  });
});
