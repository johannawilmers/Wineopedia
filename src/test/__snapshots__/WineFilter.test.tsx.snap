// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders filter component in listView > test the filter for all wines 1`] = `
<h2
  class="wineName"
>
  Tissot Vin Jaune La Vas√©e
</h2>
`;

exports[`renders filter component in listView > test the filter for all wines 2`] = `
<h2
  class="wineName"
>
  Niepoort Bioma Vintage
</h2>
`;
